lms/services/rsa_key.py

```diff
@@ -17,6 +18,44 @@ def __init__(self, db, aes_service):
         self._db = db
         self._aes_service = aes_service
 
+    def rotate(self, target_keys: int, expires_at: datetime, delete_after: datetime):
+        """
+        Expire keys older than `expires_at` and create `target_keys` new ones.
+
+        Keys have a live-cycle of:
+            active -> expired -> deleted
+
+        In flight request we signed with an active key will still be
+        verifiable as the public key endpoint query (get_all_public_jwks)
+        includes expired keys.
+
+        This method is meant to be called periodically by a celery task or similar.
+        """
+        # Delete any already expired keys that hasn't ben updated (expired) recently
+        self._db.execute(
+            delete(RSAKey).where(
+                RSAKey.expired.is_(True), RSAKey.updated <= delete_after
```

Comment:
Also include a date here to avoid deleting them in consecutive calls.

Reply:
I think I'm confused about the use of updated column here, this all seems a bit complicated. This doesn't pass the tests, but couldn't the underlying logic be something like this?

```python
        for key in self._db.query(RSAKey).all():
            if key.created >= delete_after:
                self._db.delete(key)
            elif key.created >= expires_at:
                key.expired = True
            else:
                target_keys -= 1

        for _ in range(target_keys):
            self.generate()
```